R/summarize_draws.R

```diff
@@ -0,0 +1,108 @@
+#'   summary statistic columns. The summary statistic columns have the same name
+#'   as each function specified in `summary_fun` and the quantiles are named
+#'   like 'q_`(probs * 100)`'.
+#'
+#' @details
+#' `summary_fun` correspond to names of functions in R that can take a vector of
+#' values and reduce to one summary statistic. This can also include user
+#' defined functions specified in the global environment.
+#'
+#' The `probs` argument is used to specify the probabilities to calculate sample
+#' quantiles for using the `stats::quantile()` function. The default 2.5 and
+#' 97.5 quantiles would have columns named 'q2.5' and 'q97.5' in the returned
+#' data.table.
+#'
+#' @importFrom methods existsFunction
+#' @export
+#'
+#' @examples
+#' input_dt <- data.table::data.table(location = "USA",
+#'                                    draw = 1:101,
+#'                                    value = 0:100)
+#' output_dt <- summarize_dt(dt = input_dt,
+#'                           id_cols = c("location", "draw"),
+#'                           summarize_cols = "draw",
+#'                           value_col = "value")
+summarize_dt <- function(dt,
+                         id_cols,
+                         summarize_cols,
+                         value_col,
+                         summary_fun = c("mean"),
+                         probs = c(0.025, 0.975)) {
+
+  # Validate arguments ------------------------------------------------------
+
+  # check `summarize_cols` argument
+  assertive::assert_is_character(summarize_cols)
+
+  # check `value_col` argument
+  assertthat::assert_that(assertthat::is.string(value_col))
+
+  # check `id_cols` argument
+  assertive::assert_is_character(id_cols)
+  assertthat::assert_that(all(summarize_cols %in% id_cols),
+                          msg = "`id_cols` must include `summarize_cols`")
+
+  # check `dt` argument
+  assertive::assert_is_data.table(dt)
+  assertable::assert_colnames(dt, c(id_cols, value_col), only_colnames = F,
+                              quiet = T)
+  assert_is_unique_dt(dt, id_cols)
+
+  # check `summary_fun` argument
+  assertthat::assert_that(assertive::is_character(summary_fun) |
+                            assertive::is_empty(summary_fun),
+                          all(sapply(summary_fun, methods::existsFunction)),
+                          msg = "`summary_fun` must be a charcorrespond to a defined
+                          function")
+
+  # check `probs` argument
+  assertthat::assert_that(assertive::is_numeric(probs) |
+                            assertive::is_empty(probs),
+                          all(data.table::between(probs, 0, 1)),
+                          msg = "`probs`` must be between 0 and 1
+                          (or empty/null)")
+  quantile_names <- paste0("q", probs * 100)
+
+  # Calculate summary statistics --------------------------------------------
+
+  id_cols_no_draw <- id_cols[!id_cols %in% summarize_cols]
```

Comment:
"id_cols_new" instead of "id_cols_no_draw" to remove assumption that "draw" is what we're collapsing on?